File: lode/models/Ship.py
class Ship:

    def __init__(self,name='none',type='default'):
        self.name = name
        self.type = type
        self.width = 0
        self.height = 0
        self.grid = []

    #Vytváří prázdnou mřížku
    def preCreate(self,width,height):
        self.width = width
        self.height = height
        for y in range(height):
            fr = []
            for x in range(width):
                fr.append(0)
            self.grid.append(fr)

    #Vrací počet částí lodě které jdou trefit
    def getShipParts(self):
        count = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.grid[y][x]:
                    count += 1
        return count

    #Převrací tabulku lodě po směru hodin
    def rotate(self):
        newGrid = []
        for x in range(self.width):
            fr = []
            for y in range(self.height - 1,-1,-1):
                fr.append(self.grid[y][x])
            newGrid.append(fr)
        #vyměnuje výšku a šířku lodě
        self.grid = newGrid
        mem = self.width
        self.width = self.height
        self.height = mem

    #Funkce na vytváření části loďe
    #Vylní čtverec podle zadaných souřadnic
    def fill(self,fromX,fromY,toX,toY,val = 1):
        for y in range(fromY,toY):
            for x in range(fromX,toX):
                self.grid[y][x] = val

    #Vrací data lodé v kompaktnějším formátu
    def getSaveData(self):
        large = []
        for line in self.grid:
            large.append(''.join([str(letter) for letter in line]))
        #vrací: String - naźev loďe,typ lodě,velikost x,velikost y,grid
        return '{},{},{},{},{}'.format(self.name,self.type,self.width,self.height,''.join(large))

    #argument: String - nazév loďe,typ lodě,velikost x,velikost y,grid
    #Přemění data lodě na objekt lodě
    def parseSaveData(self,data):
        arr = data.split(',')
        self.name = arr[0]
        self.type = arr[1]
        self.width = int(arr[2])
        self.height = int(arr[3])
        index = 0;
        for y in range(self.height):
            fr = []
            for x in range(self.width):
                fr.append(arr[4][index])
                index += 1
            self.grid.append(fr)
        return self



    #TESTOVACÍ
    def getGridString(self):
        out = []
        for line in self.grid:
            out.append(' '.join([str(letter) for letter in line]))
            out.append('\n')
        return ''.join(out)

File: src/objects/Ship.js
class Ship {
  width = 0;
  height = 0;
  grid = [];
  trans = false;

  constructor(name = 'default',type='none') {
    this.name = name;
    this.type = type;
  }

  preCreate(width,height){
    //Vytváří prázdnou mřížku
    this.width = width;
    this.height = height;
    for(let i = 0; i < height; i++) {
      let fragment = [];
      for(let j = 0; j < width; j++){
        fragment.push(0);
      }
      this.grid.push(fragment)
    }
  }

  rotate(){
    //Převrací tabulku lodě po směru hodin
    let newGrid = [];
    if(this.trans){
      for(let i = this.width - 1; i >= 0; i--) {
        let fragment = [];
        for(let j = this.height - 1; j >= 0; j--){
          fragment.push(this.grid[j][i]);
        }
        newGrid.push(fragment)
      }
    } else {
      for(let i = 0; i < this.width; i++) {
        let fragment = [];
        for(let j = 0; j < this.height; j++){
          fragment.push(this.grid[j][i]);
        }
        newGrid.push(fragment)
      }
    }
    //Mění výšku a šírku
    this.trans = !this.trans
    this.grid = newGrid;
    const mem = this.width
    this.width = this.height
    this.height = mem
  }

  //Funkce na vytváření části lodě
  //Vylpní čtverec podle zadaných souřadnic
  fill(fromX,fromY,toX,toY,data){
    for(let i=fromY;i<=toY;i++){
      for(let j=fromX;j<=toX;j++){
        this.grid[i][j] = data;
      }
    }

  }

  getSaveData(){
    //Vrací data tabulky v kompaktnějším formátu
    let large = []
    this.grid.forEach(i => {
      large.push(i.join(''))
    });
    return this.width + ',' + this.height + ',' + large.join('');
  }

  parseSaveData(str){
    //Přemění kompaktnější data na data lodě
    const arr = str.split(',')
    this.width = arr[0];
    this.height = arr[1];
    for(let i = 0; i < arr[1]; i++) {
      let fragment = [];
      for(let j = 0; j < arr[0]; j++){
        fragment.push(arr[2][i+j]);
      }
      this.grid.push(fragment)
    }
  }

  /* TESTOVACÍ */
  getGridString(){
    let largeStr = []
    this.grid.forEach( item => {
      largeStr.push(item.join(','))
    });
    return largeStr.join('\n')
  }


}

export default Ship;
